crates/sargon/src/profile/mfa/security_structures/security_shield_builder.rs

```diff
@@ -0,0 +1,527 @@
+use crate::prelude::*;
+
+#[derive(Debug)]
+pub struct SecurityShieldBuilder {
+    matrix_builder: RwLock<MatrixBuilder>,
+    name: RwLock<String>,
+}
+
+impl SecurityShieldBuilder {
+    pub fn new() -> Self {
+        let matrix_builder = MatrixBuilder::new();
+        let name = RwLock::new("My Shield".to_owned());
+        Self {
+            matrix_builder: RwLock::new(matrix_builder),
+            name,
+        }
+    }
+}
+
+impl SecurityShieldBuilder {
+    fn get<R>(&self, access: impl Fn(&MatrixBuilder) -> R) -> R {
+        let binding = self.matrix_builder.read().unwrap();
+        access(&binding)
+    }
+
+    // Ignores error and returns a ref to self
+    fn set<R>(&self, mut write: impl FnMut(&mut MatrixBuilder) -> R) -> &Self {
+        let mut binding = self.matrix_builder.write().expect("No poison");
+        write(&mut binding);
+        self
+    }
+
+    fn validation_for_addition_of_factor_source_by_calling(
+        &self,
+        factor_sources: Vec<FactorSourceID>,
+        call: impl Fn(
+            &MatrixBuilder,
+            &IndexSet<FactorSourceID>,
+        )
+            -> IndexSet<FactorSourceInRoleBuilderValidationStatus>,
+    ) -> Vec<FactorSourceInRoleBuilderValidationStatus> {
+        let input = factor_sources
+            .iter()
+            .map(|x| x.clone().into())
+            .collect::<IndexSet<_>>();
+        self.get(|builder| call(builder, &input))
+            .into_iter()
+            .collect_vec()
+    }
+}
+
+impl SecurityShieldBuilder {
+    fn get_factors(
+        &self,
+        access: impl Fn(&MatrixBuilder) -> &Vec<FactorSourceID>,
+    ) -> Vec<FactorSourceID> {
+        self.get(|builder| {
+            let factors = access(builder);
+            factors
+                .iter()
+                .map(|x| crate::FactorSourceID::from(*x))
+                .collect::<Vec<_>>()
+        })
+    }
+}
+
+// ====================
+// ==== GET / READ ====
+// ====================
+impl SecurityShieldBuilder {
+    pub fn get_primary_threshold(&self) -> u8 {
```

Comment:
would rename to `get_threshold` for consistency

Reply:
FIXED